### PR TITLE
CSQC: Initial support. (aka antilag 1 stuffs)

### PR DESCRIPTION
**Pull request by dsvensson**
Draft PR until #145 has been merged, and a bit more functionality has been implemented.

Minimal support for client to download and execute, and notify outdated clients that they're missing out:

```objc
void(float apilevel, string enginename, float engineversion) CSQC_Init =
{
    print("Hello World\n");
}
```

### CODE REVIEW

File: src/sv_ents.c [hunk 1]
```diff
@@ -840,6 +851,363 @@ qbool SV_EntityVisibleToClient (client_t* client, int e, byte* pvs)
...
+
+	if (sectionstarted)
+	{
+		// write index 0 to end the update (0 is never used by real entities)
+		MSG_WriteShort(msg, 0);
+	}
+	else
+	{
+		sectionstarted = 1;
+		MSG_WriteByte(msg, svc_fte_csqcentities);
```
Comment (dsvensson) [inline src/sv_ents.c:1064]: **Protocol violation vs FTE: bare, unterminated `svc_fte_csqcentities` byte — emitted every frame with no changes.**

When nothing was written (`!sectionstarted`), this emits the svc byte alone with no terminating short 0. FTE emits *nothing* when there's nothing to send, and the short-0 terminator is mandatory whenever the header byte is written (fteqw `sv_ents.c:544-545`: `if (writtenheader) MSG_WriteShort(msg, 0);`).

A conforming FTE-format parser reads a 2-byte entity index right after the svc byte and will consume the next message's bytes. Since "no csqc changes this frame" is the common case, this fires on nearly every datagram for a csqc-active client. If the current client parser tolerates or relies on this as a frame marker, that's a private dialect rather than the FTE protocol — suggest dropping the else branch entirely (FTE behavior).

File: src/sv_ents.c [hunk 2]
```diff
@@ -840,6 +851,363 @@ qbool SV_EntityVisibleToClient (client_t* client, int e, byte* pvs)
...
+			}
+
+			if (!sectionstarted)
+			{
+				MSG_WriteByte(msg, svc_fte_csqcentities);
+				sectionstarted = 1;
+			}
+
+			MSG_WriteShort(msg, (unsigned short)number);
+			msg->allowoverflow = true;
```
Comment (dsvensson) [inline src/sv_ents.c:1020]: **In-place QC writes can destroy the entire datagram on overflow.**

FTE has SendEntity write into a separate 64KB scratch buffer (`qbyte messagebuffer[MAX_OVERALLMSGLEN]`) and only copies into `msg` after checking both the QC return value and the fit (fteqw `sv_ents.c:299, 328, 396-461`). Here `csqcmsgbuffer` points at the *live* datagram with `allowoverflow = true`. If a SendEntity payload crosses the buffer's real `maxsize`, `SZ_GetSpace` **clears the whole buffer** (`SZ_Clear` in `common.c`), and the subsequent `msg->cursize = oldcursize` rollback "restores" a cursize over discarded contents — the datagram (player updates, packet entities, everything already written) is silently garbage.

Two smaller issues in the same block:
- If the QC returns false after having overflowed, `msg->overflowed` is never cleared (only the too-big path below clears it), so the datagram gets dropped later.
- An entity whose payload can never fit is retried silently every frame forever; FTE detects `cursize+5 > msg->maxsize` and prints a throttled warning (fteqw `sv_ents.c:424-425`).

Adopting FTE's scratch-buffer + commit model fixes all three at once, and gives a natural place to implement remove-on-FALSE (see comment below).

File: src/sv_ents.c [hunk 3]
```diff
@@ -840,6 +851,363 @@ qbool SV_EntityVisibleToClient (client_t* client, int e, byte* pvs)
...
+
+			if (!PR2_SendEntity(ed, client->edict, sendflags))
```
Comment (dsvensson) [inline src/sv_ents.c:1022]: **`SendEntity` returning FALSE never produces a remove.**

In FTE, if the QC declines to send but the client already has the entity, the engine emits a remove (fteqw `sv_ents.c:466-502`, gated on `PVSF_NOREMOVE`). Here the false return just rolls back and continues — the client keeps a stale ghost entity, and because `csqcentitysendflags[number]` is never cleared on this path, the QC callback is re-invoked for this entity **every frame** from now on.

This breaks the documented FTE idiom of `return FALSE` as a per-client visibility filter (`.float(entity playerent, float changedflags) SendEntity` — "0 means not to tell the client about it", which FTE converts into a remove when the client has it).

File: src/sv_ents.c [hunk 4]
```diff
@@ -969,6 +1340,19 @@ void SV_WriteEntitiesToClient (client_t *client, sizebuf_t *msg, qbool recorder)
 				continue;
 			}
 
+#ifdef FTE_PEXT_CSQC
+			if (clent && client->netchan.incoming_sequence > 5) {
```
Comment (dsvensson) [inline src/sv_ents.c:1344]: **CSQC entities are PVS-culled here; FTE never culls them — and `pvsflags` is a dead field.**

This branch is only reached for entities that passed `SV_EntityVisibleToClient` (+ `hideent`). FTE explicitly *exempts* SendEntity entities from visibility culling — "if SendEntity is set then its definitely important, even if not visible" (fteqw `sv_ents.c:3957`) — and lets the QC decide per viewer via the `playerent` parameter, honoring `PVSF_NOREMOVE` in every remove path (fteqw `sv_ents.c:348, 512`).

Consequences of culling here:
- every PVS boundary crossing generates a remove followed by a full `0xFFFFFF` resend (churn);
- global/always-visible csqc entities are impossible;
- the `pvsflags` field is mapped for mods but never read by the engine, so a mod setting `PVSF_NOREMOVE` is silently ignored.

Also: the `incoming_sequence > 5` warm-up gate has no FTE counterpart.

File: src/sv_ents.c [hunk 5]
```diff
@@ -840,6 +851,363 @@ qbool SV_EntityVisibleToClient (client_t* client, int e, byte* pvs)
...
+		{
+			// a non-csqc frame got lost... great
+			return;
+		}
+
+		// a too old frame got lost... sorry, cannot handle this
+		Con_DPrintf("CSQC entity DB: lost a frame too early to do any handling (resending ALL)...\n");
+		Con_DPrintf("Lost frame = %d\n", framenum);
+		Con_DPrintf("Entity DB = %d to %d\n", client->csqcentityframehistory[ringfirst].framenum, client->csqcentityframehistory[ringlast].framenum);
+		//EntityFrameCSQC_LostAllFrames(client);
```
Comment (dsvensson) [inline src/sv_ents.c:1132]: **The "too old to recover" path is a no-op — DarkPlaces' `LostAllFrames` was commented out.**

When a lost frame has fallen out of the ring, DP calls `EntityFrameCSQC_LostAllFrames` here to force a full resend; with it commented out, the loss is simply eaten — clients keep stale/missing csqc entities indefinitely. Setting `csqcentityframe_lastreset = -1` additionally makes *subsequent* `LostFrame` calls early-return until the next emit, so follow-up losses are unhandled too.

The ring is also effectively shallower than DP's (`NUM_CSQCENTITYDB_FRAMES = UPDATE_MASK` = 63 vs DP's 256), consumed even by duplicate-framenum entries (several emits per client packet all stamp the same `incoming_sequence`) and by empty early-return frames (the `return false` at the top of `SV_EmitCSQCUpdate` leaks the frame allocated just above it without `EntityFrameCSQC_DeallocFrame`). So "too early to handle" is not a rare case.

FTE avoids the whole class: per-outgoing-frame resend logs replayed on nack (`SV_CSQC_DroppedPacket`, fteqw `sv_ents.c:557-591`) with buffer wraparound treated as loss (`SV_ReplaceEntityFrame`, `sv_ents.c:634-648`).

File: src/sv_user.c [hunk 6]
```diff
@@ -4497,6 +4542,18 @@ void SV_ExecuteClientMessage (client_t *cl)
 
 	seq_hash = cl->netchan.incoming_sequence;
 
+#ifdef FTE_PEXT_CSQC
+	for (i = cl->csqc_latestverified + 1; i < cl->netchan.incoming_acknowledged; i++)
```
Comment (dsvensson) [inline src/sv_user.c:4546]: **Loss detection mixes incoming and outgoing sequence spaces.**

CSQC frames are stamped with `netchan.incoming_sequence` (`SV_EmitCSQCUpdate`, sv_ents.c) but verified here against `netchan.incoming_acknowledged`, which acknowledges *outgoing* sequences. FTE keys everything to `outgoing_sequence` and validates each frame slot with `frame->sequence != sequence` before pulling resend data (fteqw `sv_ents.c:301, 571-574` and `SV_AckEntityFrame`, `sv_ents.c:618-631`).

The two counters drift whenever client packet rate ≠ server frame rate (low client fps, asymmetric packet loss). Drift one way classifies every delivered frame as lost (permanent full-resend churn); the other way stamps frames below `csqc_latestverified` so they are **never scanned at all** — lost packets are then never recovered and the client keeps stale csqc state.

Also worth noting `SV_FrameAck` above is entirely commented out, so the only "delivered" signal is falling out of this loop.

File: src/sv_main.c [hunk 7]
```diff
@@ -498,6 +498,20 @@ void SV_FullClientUpdate (client_t *client, sizebuf_t *buf)
 	char info[MAX_EXT_INFO_STRING];
 	int i;
 
+#ifdef FTE_PEXT_CSQC
+	// Reki: resend all CSQC ents, for reasons. previously the initial CSQC ent states were being dropped for some delta reasons I think.
```
Comment (dsvensson) [inline src/sv_main.c:502]: This hack has no FTE counterpart, and "for reasons" is doing a lot of work here — it looks like a symptom-level patch for the two loss-recovery defects flagged in sv_ents.c/sv_user.c (sequence-space mixing + the no-op `LostAllFrames` path). `SV_FullClientUpdate` fires on connects, disconnects, frag/name changes etc., so this periodically forces full resends and masks the underlying data loss. With the loss tracking fixed (FTE-style outgoing-sequence resend logs), this block should be deletable.

File: src/pr2_cmds.c [hunk 8]
```diff
@@ -2003,7 +2014,29 @@ intptr_t PF2_FS_GetFileList(char *path, char *ext,
 #ifdef FTE_PEXT_CSQC
 intptr_t EXT_SetSendNeeded(intptr_t *args)
 {
-	PR2_RunError("SetSendNeeded not implemented yet.");
+	unsigned int subject = args[1];
```
Comment (dsvensson) [inline src/pr2_cmds.c:2017]: **Missing bounds check on `subject`.**

`subject` comes straight from the QVM args and indexes `csqcentitysendflags[subject]` (array size `MAX_EDICTS`) unchecked — an out-of-range value from the mod writes outside the array into `client_t`. FTE's equivalent checks `subject < svs.clients[to].max_net_ents` before touching the array (fteqw `pr_cmds.c:10152`, `PF_setsendneeded`).

FTE also masks the flags to 24 bits per component (matching what a QC float can round-trip); `fl` here is passed through unmasked.

File: src/sv_main.c [hunk 9]
```diff
@@ -3625,6 +3639,9 @@ void SV_InitLocal (void)
 #ifdef MVD_PEXT1_SERVERSIDEWEAPON2
 	svs.mvdprotocolextension1 |= MVD_PEXT1_SERVERSIDEWEAPON2;
 #endif
+#ifdef MVD_PEXT1_EZCSQC
+	svs.mvdprotocolextension1 |= MVD_PEXT1_EZCSQC;
```
Comment (dsvensson) [inline src/sv_main.c:3643]: `MVD_PEXT1_EZCSQC` is defined nowhere in the tree, so this block never compiles — dead code. Related gap: the csqc emit is skipped for the MVD recorder (`!recorder` in `SV_WriteEntitiesToClient`), so MVD/QTV demos contain no csqc entities at all. FTE records them: the recorder is a synthetic client with `PEXT_CSQC` + `csqcactive`, viewer = world, emitted as `svc_fte_csqcentities` inside `dem_all` (fteqw `sv_mvd.c:1637-1709`, `sv_send.c:3891-3897`), and it can embed csprogs.dat into the demo (`sv_demo_write_csqc`). Either implement the demo path or drop this `#ifdef` until it exists.

**Review by dsvensson — COMMENTED** [review]
Full review of the CSQC implementation using FTE (fteqw) as the source of truth.

**Big picture:** the core is DarkPlaces' `EntityFrameCSQC_WriteFrame`/`LostFrame` ring-buffer scheme (as `server.h` says, "code adapted from Darkplaces") wrapped in an FTE-named entry point. FTE instead tracks per-entity pending bits (`pendingcsqcbits`) plus a per-outgoing-frame resend log (`client_frame_t.resend`), keyed consistently to `outgoing_sequence` (fteqw `sv_ents.c:256-258, 557-651`). Most of what bites below traces back to that mismatch.

**What matches FTE correctly:**
- `FTE_PEXT_CSQC = 0x40000000`, `svc_fte_csqcentities = 76`, remove = entnum `| 0x8000`, short-0 batch terminator, lazy header write — all match FTE's non-REPLACEMENTDELTAS encoding (`SV_EmitDeltaEntIndex`, fteqw `sv_ents.c:279-294, 544-545`). `MAX_EDICTS = 2048` stays under the 15-bit limit so the missing `0x4000` big-index form is fine.
- csprogs announcement is byte-for-byte FTE: MD4 checksum, `*csprogs`/`*csprogssize` as `"0x%x"`, `sv_csqc_progname` cvar, warning-only for legacy clients (FTE also only warns). Only cosmetic diff: FTE leaves `*csprogsname` empty for the default name (fteqw `sv_init.c:1157-1160`).
- `enablecsqc`/`disablecsqc` clc commands with full-resend-on-enable match FTE (`sv_user.c:6037-6056`); gating emission on `csqcactive` rather than the PEXT bit matches.
- `PR2_SendEntity` mirrors FTE's Q1QVM branch precisely — `self` = ent, `other` = viewer, single sendflags arg, nonzero return = commit (fteqw `sv_ents.c:401-405`). `MSG_CSQC = 5` matches.

**Findings without a diff anchor:**
1. **No per-client CSQC reset on map change.** FTE clears `csqcactive` on level change and reallocates/zeroes `pendingcsqcbits` (fteqw `sv_init.c:660, 922`). Here `csqcactive` is reset only on drop, and scope/sendflags/frame-ring state only on fresh connect. Across a map change an entity slot reused on the new map still carries `SCOPE_ASSUMED_EXISTING`, so its first update goes out as a *delta* with stale accumulated sendflags instead of full `0xFFFFFF` — the freshly-restarted client CSQC builds the entity from partial data. Small isolated fix, worth doing first.
2. **No `.SendFlags` field harvesting.** FTE's primary change-notification is the `.SendFlags` entity field, harvested per client per send cycle and cleared by the engine (`SV_ProcessSendFlags`/`SV_CleanupEnts`, fteqw `sv_ents.c:4433-4481`), with `setsendneeded` as the secondary API and a `.Version`/`sv.csqcentversion` legacy path that also force-resends on edict reuse (`ED_Free` bumps the version, fteqw `pr_cmds.c:313-314`). Builtin-only is a defensible choice for a QVM server, but entity-reuse safety then rests entirely on the freed edict dropping out of the scan list.
3. Minor: `sv_csqcdebug` / `svc_fte_csqcentities_sized` (svc 92, per-entity length prefixes) is unimplemented — genuinely useful while debugging exactly the desync classes flagged inline.

Recommendation on ordering: the empty-batch framing bug and the in-place-buffer issues (inline in `SV_EmitCSQCUpdate`) first — adopting FTE's scratch-buffer + commit model kills three findings at once; then the loss-recovery pair (sequence-space mixing + no-op `LostAllFrames`), which is what will bite in real matches under packet loss and would let the `SV_FullClientUpdate` resend hack be deleted.

**Comment by dsvensson** [comment]
Follow-up to the review above: this is a review of the improvements in https://github.com/dusty-qw/mvdsv (branch `al`), which reworks the CSQC support reviewed here. Same methodology — FTE (fteqw) as the source of truth. Short version: **most of the serious defects flagged in the earlier review are fixed there**, the writer core is now structurally sound, but the design has deliberately diverged from FTE, one new regression appeared at the netchan boundary, and a few gaps remain.

## Architectural note

Entity serialization now requires the mvdsv-private `MVD_PEXT1_EZCSQC` extension (gated by the `qwm_ezcsqc` cvar, advertised via an `ezcsqc` serverinfo key, `sv_init.c:526-533`) on top of `FTE_PEXT_CSQC`. A genuine FTE client connecting with `PEXT_CSQC` gets **no entity stream at all** — this is an ezQuake/KTX contract carried over FTE's svc-76 framing, not FTE-interoperable CSQC.

## Fixed relative to this PR

- **Empty-batch bare svc byte** — fixed. The native writer emits nothing when there's nothing to send; the terminator only follows a header (`sv_ents.c:1389-1393`), matching FTE. (The legacy sproj writer still emits an empty keepalive, but on its own `svc_packetsprojectiles` with an on-wire framenum *and* terminator — a complete, parseable message.)
- **In-place QC writes / datagram corruption** — fixed properly. SendEntity output goes to a scratch `entbuf[MAX_MSGLEN]` with empty-payload and overflow checks and a `bytes_needed` pre-check before copying (`sv_ents.c:1306-1369`); the comment even documents the `SZ_GetSpace`-clears-the-buffer hazard. The stuck-`overflowed`-flag path is gone. This is FTE's model.
- **Sequence-space mixing** — fixed for the native path: `csqc_framenum = netchan.outgoing_sequence` (`sv_ents.c:1193-1198`), and the ack loop (`sv_user.c:4694-4699`) compares like with like. Treating every un-acked sequence below the ack as lost matches FTE's own `SV_AckEntityFrame`.
- **Loss recovery improved beyond DarkPlaces**: `EntityFrameCSQC_LostFrame` takes `latest_received_framenum` and refuses to suppress resend bits using frames the client hasn't actually acked (`sv_ents.c:1484-1492`), and recovery sets `SCOPE_WANTUPDATE`/`WANTREMOVE` directly (`sv_ents.c:1529-1537`) so recovered updates survive outside PVS.
- **Map-change reset** — fixed: `SV_ResetClientCSQCEntityState` (`sv_init.c:40-59`) clears scope/sendflags/ring and re-anchors `csqc_latestverified`, called for all clients in `SV_SpawnServer`.
- **`EXT_SetSendNeeded` bounds** — fixed (`pr2_cmds.c:2036-2062`): subject/flags validated, client state checked, and it sets `SCOPE_WANTUPDATE` so out-of-PVS scheduling works.
- **The `SV_FullClientUpdate` "Reki: for reasons" resend hack is gone** — consistent with the underlying loss recovery actually being fixed.
- **PVS, partially**: `PVSF_IGNOREPVS` is honored in `SV_EntityVisibleToClient` (`sv_ents.c:874-876`), and pending SendEntity updates survive absence from the PVS walk (`sv_ents.c:1225-1232, 1258-1260`).

## Still missing or broken there

1. **New regression: the reliable-backlog reservation was dropped — silent, unrecoverable csqc loss.** This PR passes `msg->maxsize - (netchan.message.cursize + 30)` to the emitter; the branch passes bare `msg->maxsize` (`sv_ents.c:1778`). `Netchan_Transmit` silently drops the entire unreliable datagram when the reliable part leaves no room (`net_chan.c:295-297`), and with `MAX_MSGLEN == MAX_DATAGRAM == 1450` that happens whenever reliable payload is in flight alongside a near-full datagram (downloads, stats, prints). The packet still arrives (reliable part), the client acks the sequence, so the ack-driven recovery concludes the csqc frame was **delivered** — those sendflags are gone permanently. This defeats the recovery machinery in exactly the congested moments it exists for.
2. **`SendEntity` returning FALSE still never emits a remove.** FTE converts a declined send into a remove when the client already has the entity (fteqw `sv_ents.c:466-502`). The branch clears the pending flags and moves on (`sv_ents.c:1330-1335`) — no more every-frame respam (improvement over this PR), but a client that has the entity keeps a stale ghost forever.
3. **`LostAllFrames` is still commented out** (`sv_ents.c:1463`). A loss older than the 63-entry ring is silently eaten, and `csqcentityframe_lastreset = -1` still poisons subsequent recovery until the next emit. Mitigating factor: empty frames are now dealloc'd immediately, so the ring covers the last 63 *data-carrying* frames — but the fallback remains a no-op, where FTE proactively treats overwritten history as dropped (`SV_ReplaceEntityFrame`).
4. **The legacy simple-projectile path's loss recovery is dead code.** `SV_SimpleProjectileWriteFrame_Sproj` stamps frames with a private counter (`csqc_framenum++`, reset to 0 at map spawn, offset from netchan by every pre-spawn transmit), but the shared ack loop compares those against `netchan.incoming_acknowledged`. The `framenum <= csqc_framenum` gate in `SV_FrameLost` is essentially never true, so the loop breaks immediately and `csqc_latestverified` fast-forwards — the ring is maintained but never consulted. Only the native EZCSQC path (outgoing_sequence-keyed) actually recovers. Sharing one `csqc_framenum` field between two numbering schemes is the root cause.
5. **`enablecsqc` no longer forces a resend** — a regression vs both this PR and FTE (fteqw `sv_user.c:6046-6049` marks all PRESENT ents for full resend). `SV_EnableClientsCSQC` now only sets flags (`sv_user.c:3164-3168`), so a mid-map CSQC restart on the client leaves all `ASSUMED_EXISTING` entities permanently stale. Related: `ezcsqc_ready` isn't cleared by the map-spawn reset, only by enable/disable/drop.
6. **`csqcactive` is force-enabled server-side.** `Cmd_New_f` sets it for any `PEXT_CSQC` client ("ENABLING CSQC FOR YOU! YOU'RE WELCOME", `sv_user.c:398-410`). FTE is strictly client-initiated — the client sends `enablecsqc` only after its CSQC actually initialized. Harmless for the EZCSQC gate (which also requires the extension bit), but it inverts the handshake and spams non-EZCSQC clients.
7. **PVS semantics still aren't FTE's.** FTE never culls SendEntity entities — the QC decides per-viewer. The branch default is still cull-by-PVS; `PVSF_IGNOREPVS` is an opt-out, but `PVSF_NOREMOVE` is still ignored, and the `WANTUPDATE` preservation only protects *pending* updates — an out-of-PVS entity whose update was just delivered gets a remove on the very next frame (`sv_ents.c:1258-1265`).
8. **MVD/QTV demos still contain no CSQC entities.** Every emit path is `!recorder`. FTE records them via the synthetic recorder client (`PEXT_CSQC` + `csqcactive`, viewer = world, `dem_all`) and can embed csprogs.dat into the demo (`sv_demo_write_csqc`). For MVDSV, demos/QTV silently lacking the csqc layer is arguably the biggest functional hole left.
9. **Still absent, as before:** `sv_csqcdebug`/svc-92 sized variant (would have been useful for debugging items 1–4); `.SendFlags` field harvesting and the `.Version`/`csqcentversion` edict-reuse belt-and-braces; csqc-adjacent stat plumbing (256-stat range, string stats). Cosmetic: `*csprogsname` always set; a too-big update head-of-line-blocks the stream by design (FTE warns and tries smaller entities); the per-frame overflow `Con_Printf`s are unthrottled.

## Bottom line

The branch takes the writer core from "will corrupt datagrams and desync under loss" to a sound design — scratch-buffer commit, outgoing-sequence recovery, bounds checks, map resets. What's left: one sharp regression (item 1 — small fix, restore the reserved headroom), two FTE-semantic gaps that will surface as stale ghosts (items 2 and 5), the dead sproj recovery (item 4), and the strategic facts that this is now a private EZCSQC dialect rather than FTE-compatible CSQC, with demos/QTV still blind to it (item 8).
